nx/lib/nx/binary_backend/matrix.ex

```diff
@@ -174,44 +174,47 @@ defmodule Nx.BinaryBackend.Matrix do
   defp do_ts([], [], _idx, acc), do: acc
 
   def qr(input_data, input_type, input_shape, output_type, m, k, n, opts) do
-    {_, input_num_bits} = input_type
-
     mode = opts[:mode]
     eps = opts[:eps]
 
-    r_matrix =
-      if mode == :reduced do
-        # Since we want the first k rows of r, we can
-        # just slice the binary by taking the first
-        # n * k * output_type_num_bits bits from the binary.
-        # Trick for r = tensor[[0..(k - 1), 0..(n - 1)]]
-        slice_size = n * k * input_num_bits
-        <<r_bin::bitstring-size(slice_size), _::bitstring>> = input_data
-        binary_to_matrix(r_bin, input_type, {k, n})
+    {q_matrix, r_matrix} =
+      input_data
+      |> binary_to_matrix(input_type, input_shape)
+      |> qr_decomposition(m, n, eps)
+
+    {q_matrix, r_matrix} =
+      if mode == :reduced and m - k > 0 do
+        # Remove unnecessary columns (rows) from the matrix Q (R)
+        q_matrix =
+          q_matrix
+          |> transpose_matrix()
+          |> Enum.drop(k - m)
+          |> transpose_matrix()
```

Comment:
```suggestion
        q_matrix = get_matrix_columns(q_matrix, 0..(k - 1))
```

This should work (aside from possible off-by-one issues)